### PR TITLE
Subcircuit designs

### PR DESCRIPTION
**Pull request by gmalinve**
In this PR:
- New methods have been implemented: 
1. To add a subcircuit
2. to duplicate a subcircuit
3. to push-down and pop-up from a subcircuit
- A new example has been added
- Unit tests have been added

### CODE REVIEW

File: examples/03-Circuit/Circuit_Subcircuit_Example.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+# ~~~~~~~~~~~~~~~~~~~~~~~
+# This examples launches AEDT 2022R1 in graphical mode.
+
+from pyaedt import Circuit
+
+circuit = Circuit(specified_version="2022.1")
+
+###############################################################################
+# Add new subcircuit
+# ~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (maxcapodi78) [inline examples/03-Circuit/Circuit_Subcircuit_Example.py:19]: the symbol below the title should be of the same lenght of the title.
```suggestion
# ~~~~~~~~~~~~~~~~~~
```

File: examples/03-Circuit/Circuit_Subcircuit_Example.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+r1 = circuit.modeler.schematic.create_resistor(value="R_val")
+l1 = circuit.modeler.schematic.create_inductor(value="L_val")
+c1 = circuit.modeler.schematic.create_capacitor(value="C_val")
+p2 = circuit.modeler.schematic.create_interface_port("Out")
+circuit.modeler.schematic.connect_components_in_series([p1, r1, l1, c1, p2])
+circuit.pop_up()
+
+###############################################################################
+# Subcircuit duplication
+# ~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (maxcapodi78) [inline examples/03-Circuit/Circuit_Subcircuit_Example.py:49]:
```suggestion
# ~~~~~~~~~~~~~~~~~~~~~~
```

File: examples/03-Circuit/Circuit_Subcircuit_Example.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+# The formerly created subcircuit is duplicated and new parameter value is set.
+
+new_comp = circuit.modeler.schematic.duplicate(subcircuit_name, [0.0512, 0])
+new_comp.parameters["R_val"] = "75ohm"
+
+circuit.release_desktop(False, False)
```
Comment (maxcapodi78) [inline examples/03-Circuit/Circuit_Subcircuit_Example.py:55]:
```suggestion
circuit.release_desktop(True, True)
```

File: pyaedt/modeler/PrimitivesNexxim.py [hunk 4]
```diff
@@ -79,6 +80,94 @@ def components_catalog(self):
             self._components_catalog = ComponentCatalog(self)
         return self._components_catalog
 
+    @pyaedt_function_handler()
+    def add_new_subcircuit(self, location=None, angle=0, name=None, nested_subcircuit_id=None):
```
Comment (maxcapodi78) [inline pyaedt/modeler/PrimitivesNexxim.py:84]: maybe we should rename it create_subcircuit

**Review by maxcapodi78 — COMMENTED** [review]
Overall looks, good. just few comments

**Review by maxcapodi78 — APPROVED** [review]
LGTM

**Review by maxcapodi78 — APPROVED** [review]

File: _unittest/test_21_Circuit.py [hunk 5]
```diff
@@ -355,3 +355,44 @@ def test_29_create_circuit_from_spice(self):
...
+    def test_30_create_subcircuit(self):
+        subcircuit = self.aedtapp.modeler.schematic.create_subcircuit(location=[0.0, 0.0], angle=0)
+        assert type(subcircuit.location) is list
+        assert type(subcircuit.id) is int
+        assert subcircuit.component_info
+        assert subcircuit.location[0] == "0.0mil"
+        assert subcircuit.location[1] == "0.0mil"
+        assert subcircuit.angle == 0.0
+
+    @pytest.mark.skipif(config["NonGraphical"], reason="Duplicate doesn't work in non-graphical.")
```
Comment (PipKat) [inline _unittest/test_21_Circuit.py:368]:
```suggestion
    @pytest.mark.skipif(config["NonGraphical"], reason="Duplicate doesn't work in non-graphical mode.")
```

File: examples/03-Circuit/Circuit_Subcircuit_Example.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
+"""
+Circuit: Schematic Creation and Analysis
+----------------------------------------
+This example shows how you can use PyAEDT to add a subcircuit to a Circuit design.
+ Push down into the child subcircuit and Pop up to the parent design.
```
Comment (PipKat) [inline examples/03-Circuit/Circuit_Subcircuit_Example.py:5]:
```suggestion
 Push down into the child subcircuit and pop up to the parent design.
```

Comment (PipKat) [reply examples/03-Circuit/Circuit_Subcircuit_Example.py:5]: "pop us" sounds a bit like slang?

File: examples/03-Circuit/Circuit_Subcircuit_Example.py [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+# Then it pushes down into the child subcircuit.
+
+subcircuit = circuit.modeler.schematic.create_subcircuit([0.0, 0.0])
+subcircuit_name = subcircuit.composed_name
+circuit.push_down(subcircuit)
+
+###############################################################################
+# Subcircuit parameterization
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~
+# This method parameterizes the subcircuit and adds a resistor, inductor
```
Comment (PipKat) [inline examples/03-Circuit/Circuit_Subcircuit_Example.py:31]:
```suggestion
# This method parameterizes the subcircuit and adds a resistor, inductor,
```

File: examples/03-Circuit/Circuit_Subcircuit_Example.py [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+l1 = circuit.modeler.schematic.create_inductor(value="L_val")
+c1 = circuit.modeler.schematic.create_capacitor(value="C_val")
+p2 = circuit.modeler.schematic.create_interface_port("Out")
+circuit.modeler.schematic.connect_components_in_series([p1, r1, l1, c1, p2])
+circuit.pop_up()
+
+###############################################################################
+# Subcircuit duplication
+# ~~~~~~~~~~~~~~~~~~~~~~
+# The formerly created subcircuit is duplicated and new parameter value is set.
```
Comment (PipKat) [inline examples/03-Circuit/Circuit_Subcircuit_Example.py:50]:
```suggestion
# The formerly created subcircuit is duplicated, and a new parameter value is set.
```

File: pyaedt/application/AnalysisNexxim.py [hunk 9]
```diff
@@ -52,6 +53,55 @@ def __init__(
...
+            parent_name = self.odesign.GetName().split(";")[1].split("/")[0]
+            self.oproject.SetActiveDesign(parent_name)
+            self.__init__(projectname=self.project_name, designname=parent_name)
+        except:
+            return False
+        return True
+
     @property
     def post(self):
         """PostProcessor.
```
Comment (PipKat) [inline pyaedt/application/AnalysisNexxim.py:107]:
```suggestion
        """Postprocessor.
```

File: pyaedt/application/Variables.py [hunk 10]
```diff
@@ -729,7 +730,12 @@ def set_variable(
             Whether to overwrite an existing value for the design
             property or project variable. The default is ``False``, in
             which case this method is ignored.
-
+        postprocessing : bool, optional
+            Whether to define a Post Processing Variable.
```
Comment (PipKat) [inline pyaedt/application/Variables.py:734]:
```suggestion
            Whether to define a postprocessing variable.
```

File: pyaedt/application/Variables.py [hunk 11]
```diff
@@ -729,7 +730,12 @@ def set_variable(
...
+             The default is ``False``, in which case the variable will not be defined as post processing.
```
Comment (PipKat) [inline pyaedt/application/Variables.py:735]:
```suggestion
             The default is ``False``, in which case the variable is not used in postprocessing.
```

File: pyaedt/application/Variables.py [hunk 12]
```diff
@@ -729,7 +730,12 @@ def set_variable(
...
+        circuit_parameter : bool, optional
+            Whether to define a Parameter in Circuit Design or local Parameter.
```
Comment (PipKat) [inline pyaedt/application/Variables.py:737]:
```suggestion
            Whether to define a parameter in a circuit design or a local parameter.
```

File: pyaedt/application/Variables.py [hunk 13]
```diff
@@ -729,7 +730,12 @@ def set_variable(
...
+             The default is ``True``, in which case a circuit variable is created as parameter default.
```
Comment (PipKat) [inline pyaedt/application/Variables.py:738]:
```suggestion
             The default is ``True``, in which case a circuit variable is created as a parameter default.
```

File: pyaedt/modeler/Object3d.py [hunk 14]
```diff
@@ -2858,6 +2862,28 @@ def parameters(self):
         self._parameters = ComponentParameters(self, tab, _parameters)
         return self._parameters
 
+    @property
+    def component_info(self):
+        """Component Parameters.
```
Comment (PipKat) [inline pyaedt/modeler/Object3d.py:2867]:
```suggestion
        """Component parameters.
```

File: pyaedt/modeler/Object3d.py [hunk 15]
```diff
@@ -2858,6 +2862,28 @@ def parameters(self):
...
+
+        for j in proparray:
+            propval = _retry_ntimes(10, self.m_Editor.GetPropertyValue, tab, self.composed_name, j)
+            _component_info[j] = propval
+        self._component_info = ComponentParameters(self, tab, _component_info)
+        return self._component_info
+
     @property
     def pins(self):
         """Pins of component.
```
Comment (PipKat) [inline pyaedt/modeler/Object3d.py:2889]:
```suggestion
        """Pins of the component.
```

Comment (PipKat) [reply pyaedt/modeler/Object3d.py:2889]: Should it be "Pins in the component" instead?

File: pyaedt/modeler/PrimitivesNexxim.py [hunk 16]
```diff
@@ -79,6 +80,106 @@ def components_catalog(self):
...
+        """Add a new Circuit subcircuit to the design.
+
+        Parameters
+        ----------
+        location : list of float, optional
+            Position on the X axis and Y axis.
+        angle : float, optional
+            Angle rotation in degrees. The default is ``0``.
+        name : str, optional
+            Name of the Design. The default is ``None``, in which case
```
Comment (PipKat) [inline pyaedt/modeler/PrimitivesNexxim.py:94]:
```suggestion
            Name of the design. The default is ``None``, in which case
```

File: pyaedt/modeler/PrimitivesNexxim.py [hunk 17]
```diff
@@ -79,6 +80,106 @@ def components_catalog(self):
...
+            a unique name will be generated.
```
Comment (PipKat) [inline pyaedt/modeler/PrimitivesNexxim.py:95]:
```suggestion
            a unique name is generated.
```

File: pyaedt/modeler/PrimitivesNexxim.py [hunk 18]
```diff
@@ -79,6 +80,106 @@ def components_catalog(self):
...
+        nested_subcircuit_id : str, optional
+            Id of nested subcircuit.
+            Example `"U1"`.
+
+        Returns
+        -------
+        :class:`pyaedt.modeler.Object3d.CircuitComponent`
+            Circuit Component Object when successful or False when fails.
```
Comment (PipKat) [inline pyaedt/modeler/PrimitivesNexxim.py:103]:
```suggestion
            Circuit Component Object when successful or ``False`` when failed.
```

File: pyaedt/modeler/PrimitivesNexxim.py [hunk 19]
```diff
@@ -79,6 +80,106 @@ def components_catalog(self):
...
+        Parameters
+        ----------
+        component : class:`pyaedt.modeler.Object3d.CircuitComponent` Circuit Component Object
+            Component to duplicate.
+        location : list of float, optional
+            Position on the X axis and Y axis.
+        angle : float, optional
+            Angle rotation in degrees. The default is ``0``.
+        flip : bool, optional
+            Whether the component should be flipped or not. Default value is False.
```
Comment (PipKat) [inline pyaedt/modeler/PrimitivesNexxim.py:159]:
```suggestion
            Whether the component should be flipped. The default value is ``False``.
```

File: pyaedt/modeler/PrimitivesNexxim.py [hunk 20]
```diff
@@ -79,6 +80,106 @@ def components_catalog(self):
...
+
+        Returns
+        -------
+        :class:`pyaedt.modeler.Object3d.CircuitComponent` Circuit Component Object
+        when successful or False when fails.
```
Comment (PipKat) [inline pyaedt/modeler/PrimitivesNexxim.py:164]:
```suggestion
        when successful or ``False`` when failed.
```

**Review by MaxJPRey — APPROVED** [review]

File: pyaedt/modeler/PrimitivesNexxim.py [hunk 21]
```diff
@@ -79,6 +80,106 @@ def components_catalog(self):
...
+        --------
+        >>> from pyaedt import Circuit
+        >>> cir = Circuit()
+        """
+        if not name:
+            name = generate_unique_name("Circuit")
+
+        if nested_subcircuit_id:
+            parent_name = "{}:{}:{}".format(
+                self._app.design_name.split("/")[0], nested_subcircuit_id, os.urandom(1, 10000)
```
Comment (MaxJPRey) [inline pyaedt/modeler/PrimitivesNexxim.py:115]: urandom only tke one argument. I am assuming here a low int would do it as it for example:
```suggestion
                self._app.design_name.split("/")[0], nested_subcircuit_id, os.urandom(5)
```

File: pyaedt/modeler/PrimitivesNexxim.py [hunk 22]
```diff
@@ -79,6 +80,106 @@ def components_catalog(self):
...
+            )
+        else:
+            parent_name = "{}:{}".format(self._app.design_name.split("/")[0], ":U" + str(os.urandom(1, 10000)))
```
Comment (MaxJPRey) [inline pyaedt/modeler/PrimitivesNexxim.py:118]: urandom only tke one argument. I am assuming here a low int would do it as it for example:
```suggestion
                self._app.design_name.split("/")[0], nested_subcircuit_id, os.urandom(5)
```

File: pyaedt/modeler/PrimitivesNexxim.py [hunk 23]
```diff
@@ -79,6 +80,106 @@ def components_catalog(self):
...
+        self._modeler.oeditor.Copy(["NAME:Selections", "Selections:=", comp_names])
+        location = self._get_location(location)
+        self._modeler.oeditor.Paste(
+            ["NAME:Attributes", "Page:=", 1, "X:=", location[0], "Y:=", location[1], "Angle:=", angle, "Flip:=", flip]
+        )
+        ids = [id for id in list(self.components.keys())]
+        self.refresh_all_ids()
+        id_new = [id for id in list(self.components.keys()) if id not in ids]
+        if id_new:
+            return self.components[id_new[0]]
```
Comment (MaxJPRey) [inline pyaedt/modeler/PrimitivesNexxim.py:180]:
```suggestion
        new_ids = [id for id in list(self.components.keys()) if id not in ids]
        if new_ids:
            return self.components[new_ids[0]]
```

**Pull request merged**